File: examples/05_velocity.py
"""
Description:
This script showcases the use of the Ruckig Online Trajectory 
Generation (OTG) library to compute real-time, time-optimal 
motion trajectories for a multi-degree-of-freedom (DoF) system. 
By iteratively updating kinematic states within a control loop, 
Ruckig ensures that position, velocity, acceleration, and jerk 
constraints are respected at every step, producing smooth and 
feasible motion profiles.

Use Case
-------------------
This pattern is typical in robotics (e.g., robot arms, CNC machines, pick-and-place systems),
automated camera gimbals, or any application requiring smooth, time‑optimal motion
between states under strict kinematic limits. By running the Ruckig OTG (Online Trajectory
Generation) inside a control loop, you can continuously update commands to a motor driver
or controller, ensuring adherence to acceleration and jerk constraints for safe, precise motion.

Key Features
-----------------
  • Degrees of Freedom (DoFs): 3 independent motion axes  
  • Control Cycle: 10 ms update interval (0.01 s)  
  • Control Interface: Velocity-based control (specifying target velocities and accelerations)  
  • State Specification:  
      – current_position, current_velocity, current_acceleration  
      – target_velocity, target_acceleration  
  • Kinematic Constraints:  
      – max_acceleration per axis  
      – max_jerk per axis  
  • Ruckig API Calls:  
      – `Ruckig(...)` to create an OTG instance  
      – `InputParameter` / `OutputParameter` for passing states  
      – `otg.update(...)` to compute next trajectory segment  
      – `out.pass_to_input(inp)` to feed the output back as the next input  
  • Performance Metrics:  
      – `calculation_duration` for per-step compute time  
      – `trajectory.duration` for total motion time  
  • Data Collection: accumulating `out_list` for post‑processing or plotting  
  • Optional Plotting: using an external Plotter to save a PDF of the full trajectory  

Functionality
------------------
- Creation of a 3‑DoF OTG with 10 ms control cycle  
- Velocity‑control interface selection  
- Definition of current and target kinematic states  
- Enforcement of per‑axis acceleration and jerk limits  
- Iterative update loop generating discrete trajectory points  
- Collection of output samples for downstream visualization  
- Reporting of compute time and trajectory duration  
"""

from copy import copy
from ruckig import InputParameter, OutputParameter, Result, Ruckig, ControlInterface

if __name__ == '__main__':
    # Create instances: the Ruckig OTG as well as input and output parameters
    otg = Ruckig(3, 0.01)  # DoFs, control cycle
    inp = InputParameter(3)
    out = OutputParameter(3)

    inp.control_interface = ControlInterface.Velocity

    inp.current_position = [0.0, 0.0, 0.5]
    inp.current_velocity = [3.0, -2.2, -0.5]
    inp.current_acceleration = [0.0, 2.5, -0.5]

    inp.target_velocity = [0.0, -0.5, -1.5]
    inp.target_acceleration = [0.0, 0.0, 0.5]

    inp.max_acceleration = [3.0, 2.0, 1.0]
    inp.max_jerk = [6.0, 6.0, 4.0]

    print('\t'.join(['t'] + [str(i) for i in range(otg.degrees_of_freedom)]))

    # Generate the trajectory within the control loop
    first_output, out_list = None, []
    res = Result.Working
    while res == Result.Working:
        res = otg.update(inp, out)

        print('\t'.join([f'{out.time:0.3f}'] + [f'{p:0.3f}' for p in out.new_position]))
        out_list.append(copy(out))

        out.pass_to_input(inp)

        if not first_output:
            first_output = copy(out)

    print(f'Calculation duration: {first_output.calculation_duration:0.1f} [µs]')
    print(f'Trajectory duration: {first_output.trajectory.duration:0.4f} [s]')

    # Plot the trajectory
    from pathlib import Path
    from plotter import Plotter

    project_path = Path(__file__).parent.parent.absolute()
    Plotter.plot_trajectory(project_path / 'examples' / '05_trajectory.pdf', otg, inp, out_list, plot_jerk=False)

File: examples/05_velocity.cpp
/*
 * Description:
 * -----------------------
 * This program demonstrates how to generate a time‐optimal, jerk‐limited trajectory
 * for a system with three degrees of freedom (e.g., a 3‐axis robot arm or a 3‐joint
 * mechanism) using the Ruckig online trajectory generator library. It repeatedly
 * computes the next state (position, velocity, acceleration) at each control cycle
 * until the motion constraint (target velocity and acceleration) is reached.
 *
 * In each cycle:
 *   1. The current state and target constraints are provided to the OTG.
 *   2. The OTG computes the new state that respects maximum acceleration and jerk limits.
 *   3. The new state is output for feed‐forwarding to the motion controller.
 *   4. The new state becomes the "current" state for the next cycle.
 *
 * Use Case:
 * -------------------
 * In industrial robotics or CNC machines, smooth and precise motion is critical.
 * This code can be used in a real‐time control loop to:
 *   • Decelerate or accelerate robot joints safely without exceeding mechanical limits.
 *   • Plan multi‐axis coordinated moves under velocity, acceleration, and jerk constraints.
 *   • Ensure smooth transitions to avoid vibrations or mechanical stress.
 *
 * Key Features:
 * ------------------
 * • Ruckig<3>: Instantiates a 3‐DOF online trajectory generator with a 10 ms control cycle.
 * • InputParameter<3> / OutputParameter<3>: Data structures to hold current and output states.
 * • ControlInterface::Velocity: Specifies that motion planning is based on achieving target velocities.
 * • input.current_* & input.target_*: Setting initial and goal states for position, velocity, acceleration.
 * • input.max_acceleration / input.max_jerk: Defining actuator limits for safe motion.
 * • otg.update(...): Computes the next trajectory point; returns Result::Working while motion is ongoing.
 * • output.pass_to_input(input): Transfers newly computed state to become the next cycle’s input.
 * • output.trajectory.get_duration(): Queries the total duration of the planned trajectory.
 */

#include <iostream>
#include <ruckig/ruckig.hpp>

using namespace ruckig;

int main() {
    // Create instances: the Ruckig OTG as well as input and output parameters
    Ruckig<3> otg(0.01);  // control cycle of 10 ms
    InputParameter<3> input;
    OutputParameter<3> output;

    // Set input parameters and velocity control interface
    input.control_interface = ControlInterface::Velocity;

    input.current_position = {0.0, 0.0, 0.5};
    input.current_velocity = {3.0, -2.2, -0.5};
    input.current_acceleration = {0.0, 2.5, -0.5};

    input.target_velocity = {0.0, -0.5, -1.5};
    input.target_acceleration = {0.0, 0.0, 0.5};

    input.max_acceleration = {3.0, 2.0, 1.0};
    input.max_jerk = {6.0, 6.0, 4.0};

    // Generate the trajectory within the control loop
    std::cout << "t | position" << std::endl;
    while (otg.update(input, output) == Result::Working) {
        std::cout << output.time << " | " << join(output.new_position) << std::endl;

        // Feed the output back as the next input state
        output.pass_to_input(input);
    }

    std::cout << "Trajectory duration: "
              << output.trajectory.get_duration()
              << " [s]." << std::endl;
}
